Assignment 4/src/FractionalWrapper.sol

```diff
@@ -0,0 +1,318 @@
+    ///     function exists purely to comply with `EIP4626`.
+    function asset() external override view returns (address) {
+        return address(token);
+    }
+
+    /// @notice The name of the wrapper token.
+    string public constant override name = "FractionalWrapper";
+    /// @notice The symbol of the wrapper token.
+    string public constant override symbol = "FWRAP";
+    /// @notice The number of decimals of the wrapper token.
+    uint8 public constant override decimals = 18;
+
+    /// @notice The fraction at which the wrapped tokens are minted for the
+    ///     underlying.
+    /// @dev Denoted in increments of 1e-27. `fractionDenominator` gives the
+    //      unit (corresponding to a fraction of 1:1).
+    ///
+    ///     Note: This conversion happens in terms of the smallest unit of the
+    ///     token. The wrapping token has a fixed number of decimals that may
+    ///     be different to the number of decimals of the underlying. If that
+    ///     is the case, an unintended conversion may happen. For example,
+    ///     wrapping a token with 0 decimals into a wrapper with 18 decimals
+    ///     will effectively correspond to a fraction of
+    ///     `1 / 1e-18 * fractionDenominator`.
+    uint256 public immutable fraction;
+    /// @notice The unit corresponding to a fraction of 1 in terms of the
+    ///     smallest token unit. See also `fraction`.
+    uint256 constant fractionDenominator = 10**27;
+
+    /// @notice The total amount of wrapped tokens in existence.
+    uint256 public override totalSupply = 0;
+
+    /// @notice The total amount of assets managed by the vault.
+    /// @dev This simply calls the balance of this address. This means that not
+    ///     all assets may be redeemable by an account: we could have received
+    ///     a generous gift. These would still add up to the total.
+    /// @return totalManagedAssets The assets managed by this contract.
+    function totalAssets() external override view returns (uint256 totalManagedAssets) {
+        return token.balanceOf(address(this));
+    }
+
+    /// @notice The current allowance by the owner to spend by the spender.
+    ///     Access as follows:
+    ///     - owner: The owner of the tokens. Note: The allowance may be higher
+    ///         than their current balance.
+    ///     - spender: The spender of the tokens.
+    mapping(address => mapping(address => uint256)) public override allowance;
+
+    /// @notice The balance of addresses in terms of wrapper tokens.
+    mapping(address => uint256) public override balanceOf;
+
+    /// @notice Create a new wrapper.
+    /// @param _token The token to use as collateral.
+    /// @param _fraction The fraction at which to mint/withdraw the tokens.
+    /// @dev See the caveats on `fraction` for how to use the `_fraction`
+    ///     parameter.
+    constructor(IERC20 _token, uint256 _fraction) {
+        token = _token;
+        fraction = _fraction;
```

Comment:
The assignment might not have been very clear. The exchange rate can be also be set by the contract owner _after_ the contract is deployed.